src/rest_framework_dso/fields.py

```diff
@@ -9,21 +9,234 @@
+                return ALLOW_ALL_FIELDS_TO_DISPLAY
+
+    def get_allow_list(self, valid_names: set[str]) -> tuple[set[str], set[str]]:
+        """Find out which fields should be included.
+        This transforms the include/exclude behavior
+        into a positive list of fields that should be kept.
+        """
+        # Detect any invalid names
+        if self._allow_all:
+            return valid_names, set()
+        elif self._excludes:
+            fields_to_keep = valid_names - self._excludes
+            invalid_fields = self._excludes - valid_names
+        elif not self._includes and self._children:
+            # Only sub-objects are limited, so the 'include' should not be used
+            # for limiting the fields at all.
+            return valid_names, set()
+        else:
+            # The includes list is leading, even when its empty.
+            fields_to_keep = set(self._includes)  # copy to avoid tampering
+            invalid_fields = fields_to_keep - valid_names
+
+        # To exclude a sublevel attribute, the parent must always be included:
+        fields_to_keep.update(self._children.keys())
+        return fields_to_keep, invalid_fields
+
+    def apply(
+        self,
+        fields: dict[str, serializers.Field],
+        valid_names: Iterable[str],
+        always_keep: Iterable[str],
+    ) -> dict[str, serializers.Field]:
+        """Reduce the fields from a serializer based on the current context.
+        This returns a new dictionary that can be assigned to `serializer.fields`.
+        """
+        if self._allow_all:
+            return fields
+
+        # Split into fields to include and fields to omit (-fieldname).
+        valid_names = set(fields.keys()) | set(valid_names)
+        fields_to_keep, invalid_fields = self.get_allow_list(valid_names)
+        fields_to_keep.update(always_keep)
+
+        if invalid_fields:
+            # Some of `display_fields` are not in result.
+            names_str = f"', {self.prefix}'".join(sorted(invalid_fields))
+            raise ValidationError(
+                ngettext(
+                    "The following field name is invalid: {names}.",
+                    "The following field names are invalid: {names}.",
+                    len(invalid_fields),
+                ).format(names=f"'{self.prefix}{names_str}'"),
+                code="fields",
+            )
+
+        # Python 3.6+ dicts are already ordered, so there is no need to use DRF's OrderedDict
+        return {
+            field_name: field
+            for field_name, field in fields.items()
+            if field_name in fields_to_keep
+        }
 
-def parse_request_fields(fields: Optional[str]):
-    if not fields:
-        return None
 
-    # TODO: support nesting, perform validation
-    return fields.split(",")
+ALLOW_ALL_FIELDS_TO_DISPLAY = FieldsToDisplay()
+DENY_SUB_FIELDS_TO_DISPLAY = FieldsToDisplay()
+DENY_SUB_FIELDS_TO_DISPLAY._allow_all = False
```

Comment:
is it safe to re-use these instances when determining nested representations for different sub fields?

Reply:
good point, yes for now. The objects are not mutated themselves after init. (unless you start meddling with the returned sets perhaps)

Reply:
The are marked up as contstants, hence should be type annotated as such:

```python
ALLOW_ALL_FIELDS_TO_DISPLAY: Final = FieldsToDisplay()
DENY_SUB_FIELDS_TO_DISPLAY: Final = FieldsToDisplay()._allow_all = False
```